File: leetcode/string/443. String Compression/js/main.js
// Time: O(n)
// Space: O(1)

const compress = chars => {
  let ans = 0;
  let index = 0;

  while (index < chars.length) {
    let curChar = chars[index];
    let count = 0;

    while (index < chars.length && curChar === chars[index]) {
      index++;
      count++;
    }

    chars[ans] = curChar;
    ans++;

    if (count !== 1) {
      const countArr = count.toString().split('');

      for (let count of countArr) {
        chars[ans] = count;
        ans++;
      }
    }
  }

  return ans;
};

File: leetcode/string/443. String Compression/py/main.py
class Solution:
    # Time O(N)
    # Space O(1)
    def compress(self, s):
        n = len(s)
        i = 0
        index = 0

        while i < n:
            j = i + 1

            while j < n and s[i] == s[j]:
                j += 1

            cnt = j - i

            if cnt > 1:
                s[index] = s[i]
                index += 1
                for c in str(cnt):
                    s[index] = c
                    index += 1
            else:
                s[index] = s[i]
                index += 1

            i = j

        while len(s) > index:
            s.pop()